index.js

```diff
@@ -72,6 +72,23 @@ module.exports = {
     "simple-import-sort/imports": "error",
     "simple-import-sort/exports": "error",
     "sort-destructure-keys/sort-destructure-keys": "error",
+    /**
+     * List of sonarjs rules we turned off after release of eslint-plugin-sonarjs v2
+     * Some rules are too restrictive for common use cases,
+     * and others are already covered by other plugins used here.
+     */
+    "sonarjs/different-types-comparison": "off",
```

Comment:
(All examples from comments will be from Schumacher).

This rule appears to be helpful. Was it turned off because it overlaps with another rule? It surfaces:

```diff
diff --git a/src/helpers/numberHelpers.ts b/src/helpers/numberHelpers.ts
index 0866920f58..a40ad3e228 100644
--- a/src/helpers/numberHelpers.ts
+++ b/src/helpers/numberHelpers.ts
@@ -30,7 +30,7 @@ export const parseNumber = (value: string) => {
  * Parses a string into an integer value or null if not able to be parsed.
  */
 export const parseNumberAtMaxLength =
-  (maxLength: number) => (value: string) => {
+  (maxLength: number | undefined) => (value: string) => {
     const parsedNumber = parseNumber(value);
     if (parsedNumber === null || maxLength === undefined) {
       return parsedNumber;
```

We very often have incorrect type definitions where we are applying conditional logic that, according to the type definitions, is unnecessary, e.g. something is typed as non-nullish and we are treating it like it might not be. This rule could be a helpful one to surface such code smells automatically.

Perhaps it was turned to `off` because there are lots of existing violations? Can we set it to `warn`, with consideration for a TODO to set it to `error` in the future when there is capacity to resolve violations?

Another example where we are casting to a stronger defined type than runtime seems to anticipate:

<img width="566" alt="Screenshot 2024-09-12 at 10 33 33 AM" src="https://github.com/user-attachments/assets/07541c78-794c-4e89-b207-32e70a837356">

Reply:
The main violations were when there was an object having a union type, each with a different child name
```
type Type1 = {
   child1: string;
  }
type Type2 = {
   child2: number;
}
  ```
One common pattern we've used is to do if ((obj as Type1).child !== undefined) for example to find out at runtime what the object is, but this rule would throw errors that we're comparing objects of different types, which is the point of the comparison

Reply:
But okay turning this to a warning

Reply:
> One common pattern we've used is to do if ((obj as Type1).child !== undefined) for example to find out at runtime what the object is, but this rule would throw errors that we're comparing objects of different types, which is the point of the comparison

`'child1' in obj` should be the preferred syntax, without `as` casting (which would require a custom lint rule to forbid, something I've looked forward to).

In all these cases I think the motivation for our rules should be to guide/enforce "happy paths" and limit comparisons (like those with as casting, or testing different types without doing so in a way that is compliant with TypeScript) that do not fall into that list. Sometimes the comments on these rules can prescribe the happy paths, sometimes we'll need to have docs around "best practices," but other times we will just need devs to adjust to rules (and/or start dialogues with team members) to learn these patterns directly.